### ISSUE TITLE
[Pet ReID] 训练可部署的整宠身份特征 CNN（Open Noodle 蒸馏）

### ISSUE THREAD

**Issue by klrc** [molchip-aiml/molchip-model-zoo#22]
## 目标

训练一个可部署的猫/狗个体识别特征模型，用于家庭摄像头中的“这是哪一只宠物”，不是品种分类。

产品链路：

```text
猫狗检测/跟踪 -> 选择清晰的整宠框 -> 特征提取 -> 多视角特征库余弦匹配 -> 未知宠物拒绝
```

首个候选固定为：

- `MobileNetV3-Small`，ImageNet 预训练，允许全量微调；
- 输入 `RGB 192x192` 整宠框，不要求正脸或宠物头部检测；
- 输出 `512D float32`、L2-normalized embedding；
- 猫狗共用一个 checkpoint；
- 纯静态 CNN，不使用 Transformer、attention、data-by-data MatMul、动态卷积或动态 shape；
- 身份比较使用 cosine，不在模型内加入宠物品种分类结果。

不要提前分配 Model Zoo IID，也不要发布 Release。训练候选通过本 Issue 的精度和导出门禁后，再进入编译、量化和板端资格流程。

## Teacher

[Open Noodle pet-recognition-small](https://huggingface.co/open-noodle/pet-recognition-small) 固定为蒸馏 teacher：

- revision: `9dd4c915be29a81b116b3e30eb996c59d0e7ede0`
- license: Apache-2.0
- input: RGB `224x224`，ImageNet mean/std
- output: 512D L2-normalized embedding
- backbone: frozen DINOv2-S；teacher 只离线生成 soft embedding，不进入最终模型
- upstream ONNX SHA256: `6a5e2373ab348bed588cef4072f3914ca9c8bacde3e8d0651019e8dad86b24ba`

Teacher 已确认不适合直接作为边缘产物：INT8 embedding 偏差过大；恢复精度所需的混合精度范围又使计算成本过高。因此本任务是训练新的 CNN student，不再继续调 teacher 的 PTQ 参数。

### 训练集

1. [Dogs of the World](https://www.kaggle.com/datasets/lextoumbourou/dogs-world)，CC0
   - 使用每张图片的 metadata sidecar 提取个体身份；
   - 只保留单狗图片、同一身份至少两张图片的样本；
   - 原始数据约 30 万张，不能直接假定全部 metadata 正确，必须先去重和审计。

2. [Cat Individual Images](https://www.kaggle.com/datasets/timost1234/cat-individuals/data)，CC BY 4.0
   - 目录对应个体身份；
   - 约 13,542 张、518 只猫；
   - 过滤多猫、错误目录、损坏图及跨身份重复图。

训练、验证、测试必须按宠物身份切分，不能按图片随机切分。近重复图片、同一连拍或视频帧不能跨 split。生成固定 manifest，记录相对路径、identity、species、split、原文件 SHA256；数据图片不提交到 Model Zoo Git 仓库。

### 外部验证集

[DogFaceNet_224resized](https://zenodo.org/records/12578449)，CC BY 4.0，只用于外部验证，不参与训练或阈值选择：

- archive: `DogFaceNet_224resized.zip`
- 1,393 个身份、8,363 张 224x224 对齐狗脸
- 使用发布方提供的 `classes_train.txt` / `classes_test.txt`
- archive SHA256: `b3b335180bfd8d18b17e13601c9b0fa9c7c92bf9c18d64fe2999597f2e71f871`

不要使用 PetFace、AvitoTech animal-identification、MegaDescriptor 权重或其训练数据进行蒸馏/训练；它们当前存在非商业许可或数据来源限制。

## 数据处理

- 同一个整宠框分别 resize 为 teacher 的 `224x224` 和 student 的 `192x192`，两者使用相同的几何增广；
- RGB，ImageNet mean/std；
- 增广覆盖视角、尺度、亮度、色偏、轻度模糊、遮挡、背景变化和 collar/clothing 变化；
- 不做人脸式五点对齐，不要求宠物正脸；
- 一个 batch 按 identity 采样，每个 identity 至少两张图；猫狗需要平衡，避免 Dogs-World 规模压制猫数据；
- 固定随机种子并保存最终 split/manifest，不允许用 test identity 调参。

## 训练基线

建议先运行一个小规模可复现 pilot，通过后再扩到全部合格数据：

1. 加载 ImageNet MobileNetV3-Small；
2. 新建 512D projection + L2 normalization；
3. teacher 全程冻结并预计算 512D embedding；
4. student loss 至少包含：
   - cosine embedding distillation；
   - 基于真实 identity 的 ArcFace classification loss（分类头仅训练时使用）；
5. 先短暂冻结 backbone 训练 projection，再以较小学习率联合微调 backbone；
6. 保存 best checkpoint 的依据必须是 identity-disjoint validation，而不是训练 identity 的分类准确率。

可以调整 optimizer、学习率、loss 权重和 epoch，但必须在回执中记录；不要在本任务中展开大规模 backbone 搜索。只有 MobileNetV3-Small 明确达不到精度门槛时，才补一个 MobileNetV3-Large 对照。

## 评测与门禁

所有指标分别报告猫、狗和合计结果，并同时报告单张注册及最多 3 张多视角注册：

- closed-set Top-1、Top-5、mAP；
- open-set ROC/AUC、EER、TAR/Recall at FAR 1% 和 5%；
- same-breed hard negatives；
- teacher/student embedding cosine 与 pairwise-similarity correlation；
- unknown identity rejection threshold；
- 不同背景、姿态、遮挡和时间变化分组结果。

最低晋级门槛：

- 在同一固定 held-out split 上，student 的 Top-1 不低于 teacher 的 90%；
- student EER 不高于 teacher EER `+0.02`；
- DogFaceNet 外部验证 Top-1 不低于 teacher 的 95%，EER 不高于 teacher EER `+0.02`；
- 多视角注册不得劣于单张注册；
- 没有 identity overlap、跨 split 的 exact duplicate 或明显的近重复泄漏。

Open Noodle 公布的参考值为：Dogs-World Top-1 `0.535` / EER `0.068`，Cat Individual Images Top-1 `0.913` / EER `0.065`，DogFaceNet Top-1 `0.899` / EER `0.055`。最终比较必须使用本次固定 manifest 重新跑 teacher，不能仅复制 model card 数字。

## 交付物

- [ ] 训练代码、依赖锁定文件和完整命令
- [ ] 固定的 train/validation/test manifest 与审计报告（不包含数据图片）
- [ ] teacher embedding cache 的生成脚本及 teacher revision/SHA
- [ ] best checkpoint、checkpoint SHA256 和训练日志
- [ ] 静态 ONNX：float32 NCHW `[1,3,192,192] -> [1,512]`，opset 15 或 17
- [ ] ONNX 与 PyTorch 至少 100 张输入的逐样本 cosine/max-abs 对齐报告
- [ ] 猫、狗、合计的完整身份识别与验证指标 JSON
- [ ] 模型参数量、MACs/FLOPs、ONNX bytes 和算子统计
- [ ] 数据来源、许可证、使用方式和已知限制说明

完成上述交付后，由 Model Zoo owner 负责独立复验、量化集构造、编译、量化精度、资源和板端门禁；训练设备无需创建 IID、修改 catalog 或上传 Model Zoo Release。
